src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -103,10 +97,11 @@ public virtual void Dispose()
                 current = default(TSource);
                 state = -1;
             }
-
+            
             public IEnumerator<TSource> GetEnumerator()
             {
-                Iterator<TSource> enumerator = state == 0 && _threadId == Environment.CurrentManagedThreadId ? this : Clone();
+                if (Interlocked.CompareExchange(ref state, 1, 0) == 0) return this;
```

Comment:
@VSadov 

* If state is 0, we have to do the CAS anyway.
* If state is not 0, we are going to allocate a new reference type, which is more expensive that a compare exchange.

I don't think test-and-CAS can save anything here, and actually introduces additional work for the most common path.